File: addons/Support/Imaging.cs
using System;
using System.Collections.Generic;
using System.Text;
using System.Drawing;
using System.Drawing.Imaging;
using System.Reflection;
using System.Runtime.InteropServices;

namespace Support
{
    public class Imaging
    {
        public static Bitmap RotateImage(Image image, float angle)
        {
            double num9;
            double num10;
            double num11;
            double num12;
            if (image == null)
            {
                throw new ArgumentNullException("image");
            }
            double width = image.Width;
            double height = image.Height;
            double num3 = (angle * Math.PI) / 180;
            double d = num3;
            while (d < 0)
            {
                d += 2 * Math.PI;
            }
            if (((d >= 0) && (d < Math.PI / 2)) || ((d >= Math.PI) && (d < Math.PI*1.5)))
            {
                num9 = Math.Abs(Math.Cos(d)) * width;
                num10 = Math.Abs(Math.Sin(d)) * width;
                num11 = Math.Abs(Math.Cos(d)) * height;
                num12 = Math.Abs(Math.Sin(d)) * height;
            }
            else
            {
                num9 = Math.Abs(Math.Sin(d)) * height;
                num10 = Math.Abs(Math.Cos(d)) * height;
                num11 = Math.Abs(Math.Sin(d)) * width;
                num12 = Math.Abs(Math.Cos(d)) * width;
            }
            double a = num9 + num12;
            double num6 = num11 + num10;
            int num7 = (int)Math.Ceiling(a);
            int num8 = (int)Math.Ceiling(num6);
            Bitmap bitmap = new Bitmap(num7, num8);
            using (Graphics graphics = Graphics.FromImage(bitmap))
            {
                Point[] pointArray;
                if ((d >= 0) && (d < Math.PI/2))
                {
                    pointArray = new Point[] { new Point((int)num12, 0), new Point(num7, (int)num10), new Point(0, (int)num11) };
                }
                else if ((d >= Math.PI / 2) && (d < Math.PI))
                {
                    pointArray = new Point[] { new Point(num7, (int)num10), new Point((int)num9, num8), new Point((int)num12, 0) };
                }
                else if ((d >= Math.PI) && (d < Math.PI*1.5))
                {
                    pointArray = new Point[] { new Point((int)num9, num8), new Point(0, (int)num11), new Point(num7, (int)num10) };
                }
                else
                {
                    pointArray = new Point[] { new Point(0, (int)num11), new Point((int)num12, 0), new Point((int)num9, num8) };
                }
                graphics.DrawImage(image, pointArray);
            }
            return bitmap;
        }

        public static void BmpAddFast(Bitmap from, Bitmap addTo, Rectangle area)
        {

            BitmapData bitmapdata = addTo.LockBits(area, ImageLockMode.WriteOnly, PixelFormat.Format24bppRgb);
            byte[] destination = new byte[bitmapdata.Stride * bitmapdata.Height];
            Marshal.Copy(bitmapdata.Scan0, destination, 0, destination.Length);

            BitmapData bitmapdata2 = from.LockBits(new Rectangle(0,0,from.Width, from.Height), ImageLockMode.ReadOnly, PixelFormat.Format24bppRgb);
            byte[] source = new byte[bitmapdata2.Stride * bitmapdata2.Height];
            Marshal.Copy(bitmapdata2.Scan0, source, 0, source.Length);
            from.UnlockBits(bitmapdata2);

            int num = 0;
            for (int i = from.Height - 1; i >= 0; i--)
            {
                for (int j = 0; j < from.Width; j++)
                {
                    destination[num] = source[num++];
                    destination[num] = source[num++];
                    destination[num] = source[num++];
                }
            }
            Marshal.Copy(destination, 0, bitmapdata.Scan0, destination.Length);
            addTo.UnlockBits(bitmapdata);
        }

        public static Bitmap BmpGetRegion(Bitmap from, Rectangle area)
        {

            Color bgra = new Color();
            Bitmap addTo = new Bitmap(area.Width, area.Height, PixelFormat.Format24bppRgb);
            BitmapData bitmapdata = addTo.LockBits(new Rectangle(0,0,area.Width, area.Height), ImageLockMode.WriteOnly, PixelFormat.Format24bppRgb);
            byte[] destination = new byte[bitmapdata.Stride * bitmapdata.Height];
            Marshal.Copy(bitmapdata.Scan0, destination, 0, destination.Length);
            int num = 0;
            for (int i = area.Y; i < from.Height && i<area.Bottom; i--)
            {
                for (int j = area.X; j < from.Width && j<area.Right; j++)
                {
                    bgra = from.GetPixel(j, i);
                    destination[num++] = bgra.B;
                    destination[num++] = bgra.G;
                    destination[num++] = bgra.R;
                }
            }
            Marshal.Copy(destination, 0, bitmapdata.Scan0, destination.Length);
            addTo.UnlockBits(bitmapdata);

            return addTo;
        }
    }

    struct ColorBgra
    {
        public byte A;
        public byte R;
        public byte B;
        public byte G;
        public Color ToColor()
        {
            return Color.FromArgb(A, R, G, B);
        }
    }
}

File: Imaging.vb
Imports System.Drawing
Imports System.Drawing.Imaging
Imports System.Runtime.InteropServices

Namespace Support
    Public Class Imaging
        Public Shared Function RotateImage(ByVal img As Image, ByVal angle As Single) As Bitmap
            Dim n9, n10, n11, n12 As Double
            If img Is Nothing Then Return Nothing

            Dim width As Double = img.Width
            Dim height As Double = img.Height
            Dim n3 As Double = (angle * Math.PI) / 180
            Dim d As Double = n3
            Do While (d < 0)
                d += 2 * Math.PI
            Loop
            If (((d >= 0) AndAlso (d < Math.PI / 2)) OrElse ((d >= Math.PI) AndAlso (d < Math.PI * 1.5))) Then
                n9 = Math.Abs(Math.Cos(d)) * width
                n10 = Math.Abs(Math.Sin(d)) * width
                n11 = Math.Abs(Math.Cos(d)) * height
                n12 = Math.Abs(Math.Sin(d)) * height
            Else
                n9 = Math.Abs(Math.Sin(d)) * height
                n10 = Math.Abs(Math.Cos(d)) * height
                n11 = Math.Abs(Math.Sin(d)) * width
                n12 = Math.Abs(Math.Cos(d)) * width
            End If
            Dim a As Double = n9 + n12
            Dim n6 As Double = n11 + n10
            Dim n7 As Integer = Math.Ceiling(a)
            Dim n8 As Integer = Math.Ceiling(n6)
            Dim bmp As Bitmap = New Bitmap(n7, n8)
            Using gr As Graphics = Graphics.FromImage(bmp)
                Dim pointArray() As Point
                If ((d >= 0) AndAlso (d < Math.PI / 2)) Then
                    pointArray = New Point() {New Point(n12, 0), New Point(n7, n10), New Point(0, n11)}
                ElseIf ((d >= Math.PI / 2) AndAlso (d < Math.PI)) Then
                    pointArray = New Point() {New Point(n7, n10), New Point(n9, n8), New Point(n12, 0)}
                ElseIf ((d >= Math.PI) AndAlso (d < Math.PI * 1.5)) Then
                    pointArray = New Point() {New Point(n9, n8), New Point(0, n11), New Point(n7, n10)}
                Else
                    pointArray = New Point() {New Point(0, n11), New Point(n12, 0), New Point(n9, n8)}
                End If
                gr.DrawImage(img, pointArray)
            End Using
            Return bmp
        End Function

        Public Shared Sub BmpAddFast(ByVal from As Bitmap, ByVal addTo As Bitmap, ByVal area As Rectangle)
            Dim bmpData As BitmapData = addTo.LockBits(area, ImageLockMode.WriteOnly, PixelFormat.Format24bppRgb)
            Dim destination(bmpData.Stride * bmpData.Height) As Byte
            Marshal.Copy(bmpData.Scan0, destination, 0, destination.Length)

            Dim bmpData2 As BitmapData = from.LockBits(New Rectangle(0, 0, from.Width, from.Height), ImageLockMode.ReadOnly, PixelFormat.Format24bppRgb)
            Dim source(bmpData2.Stride * bmpData2.Height) As Byte
            Marshal.Copy(bmpData2.Scan0, source, 0, source.Length)
            from.UnlockBits(bmpData2)

            Dim cnt As Integer = 0
            For i As Integer = from.Height - 1 To 0 Step -1
                For j As Integer = 0 To from.Width - 1 Step 1
                    destination(cnt) = source(cnt)
                    cnt += 1
                    destination(cnt) = source(cnt)
                    cnt += 1
                    destination(cnt) = source(cnt)
                    cnt += 1
                Next
            Next
            Marshal.Copy(destination, 0, bmpData.Scan0, destination.Length)
            addTo.UnlockBits(bmpData)
        End Sub

        Public Shared Function BmpGetRegion(ByVal from As Bitmap, ByVal area As Rectangle) As Bitmap

            Dim bgra As New Color()
            Dim addTo As New Bitmap(area.Width, area.Height, PixelFormat.Format24bppRgb)
            Dim bmpData As BitmapData = addTo.LockBits(New Rectangle(0, 0, area.Width, area.Height), ImageLockMode.WriteOnly, PixelFormat.Format24bppRgb)
            Dim destination(bmpData.Stride * bmpData.Height) As Byte
            Marshal.Copy(bmpData.Scan0, destination, 0, destination.Length)
            Dim cnt As Integer = 0
            For i As Integer = area.Y To from.Height - 1 OrElse area.Bottom - 1 Step -1
                For j As Integer = area.X To from.Width - 1 OrElse area.Right - 1 Step 1
                    bgra = from.GetPixel(j, i)
                    destination(cnt) = bgra.B
                    cnt += 1
                    destination(cnt) = bgra.G
                    cnt += 1
                    destination(cnt) = bgra.R
                    cnt += 1
                Next
            Next
            Marshal.Copy(destination, 0, bmpData.Scan0, destination.Length)
            addTo.UnlockBits(bmpData)
            Return addTo
        End Function
    End Class
End Namespace